contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,472 @@
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_SELECTOR))
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Transfers a given amount of a PSP22 token on behalf of a specified account to another account
+        ///
+        /// Will revert if not enough allowance was given to the caller prior to executing this tx
+        fn transfer_from_tx(
+            &self,
+            token: AccountId,
+            from: AccountId,
+            to: AccountId,
+            amount: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(TRANSFER_FROM_SELECTOR))
+                        .push_arg(from)
+                        .push_arg(to)
+                        .push_arg(amount)
+                        .push_arg(vec![0x0]),
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true)) // needed for checking allowance before the actual tx
+                .returns::<Result<(), PSP22Error>>()
+                .fire()
+        }
+
+        /// Returns the amount of unused allowance that the token owner has given to the spender
+        fn allowance(
+            &self,
+            token: AccountId,
+            owner: AccountId,
+            spender: AccountId,
+        ) -> Result<Balance, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(ALLOWANCE_SELECTOR))
+                        .push_arg(owner)
+                        .push_arg(spender),
+                )
+                .returns::<Balance>()
+                .fire()
+        }
+
+        /// Returns DEX balance of a PSP22 token for an account
+        fn balance_of(&self, token: AccountId, account: AccountId) -> Result<Balance, InkEnvError> {
+            build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(token))
+                .exec_input(
+                    ExecutionInput::new(Selector::new(BALANCE_OF_SELECTOR)).push_arg(account),
+                )
+                .returns::<Balance>()
+                .fire()
+        }
+
+        fn out_given_in(
+            amount_token_in: Balance,
+            balance_token_in: Balance,
+            balance_token_out: Balance,
+            swap_fee: u128,
+        ) -> Result<Balance, DexError> {
+            let op0 = amount_token_in
+                .checked_mul(swap_fee)
+                .and_then(|result| result.checked_div(100))
```

Comment:
Since swap_fee can only be within {0,100} range, this division will always yield 0.

Reply:
For a moment I thought that maybe my mistake is caused by the fact that I'm assuming the `Balance` type to beheave like `u128`, after all - it could be some composite type capable of representing fractions (possibly via two unsigned integers) but I've checked `ink` and it seems to be just `u128` by default:
https://github.com/paritytech/ink/blob/master/crates/env/src/types.rs#L186

Do we change it anywhere?